src/main/java/demo/ReadBalancingDataSource.java

```diff
@@ -32,18 +32,25 @@ public ReadDataSource(DataSource dataSource) {
         DataSource dataSource;
         AtomicLong blacklistedUntil;
 
-        public boolean isBlackListed() {
-            long blacklistedUntilTime = this.blacklistedUntil.get();
-            
-            if (blacklistedUntilTime == -1) {
-                return false;
-            }
-            
-            if (blacklistedUntilTime > System.currentTimeMillis()) {
-                this.blacklistedUntil.set(-1);
-                return true;
-            }
-            return false ;
+        public boolean isBlacklisted() {
+            long currentBlacklistedUntil;
+            do {
+                currentBlacklistedUntil = this.blacklistedUntil.get();
+
+                if (currentBlacklistedUntil == -1) {
+                    return false;
+                }
+
+                if (currentBlacklistedUntil < blacklistClockTime()) {
+                    return true;
+                }
+            } while (blacklistedUntil.compareAndSet(currentBlacklistedUntil, -1)); // CAS loop
```

Comment:
Shouldn't this have "!" in front of it?  We'd only want to go around again if the CAS test fails (meaning someone else moved it).